axlearn/common/checkpointer.py

```diff
@@ -404,7 +404,22 @@ def restore_from_dir(
         ckpt_dir: str,
         validation: CheckpointValidationType = CheckpointValidationType.EXACT,
         concurrent_gb: int = 32,
+        dtype_cast_func: Optional[Callable[[jnp.dtype], jnp.dtype]] = None,
```

Comment:
Reading the unittest again, I'm not sure we need this new arg. It seems that user can manipulate `state` by casting one type to another instead. Would that work?

Reply:
do you mean we can move it to TensorStoreStateStorageBuilder to manipulate the state? 

https://github.pie.apple.com/foundation-models/axlearn/blob/6f5d47153067bb3186b070f370b3e34db87c862b/axlearn/common/state_builder.py#L390

Reply:
Maybe with a chain state builder? Let's discuss more offline. I'd like to understand your use case better.

Reply:
I checked the chain state builder. But I still this could be a native feature of inference runner when loading checkpoints. I made another change here as a reference here. https://github.com/apple/axlearn/pull/296. This might also explain why the inference is kind of slow and uses more memory since it has a data type cast for each iter.